gems/aws-sdk-core/lib/aws-sdk-core/plugins/client_metrics_plugin.rb

```diff
@@ -0,0 +1,191 @@
+require 'date'
+
+module Aws
+  module Plugins
+    class ClientMetricsPlugin < Seahorse::Client::Plugin
+
+      option(:client_side_monitoring,
+        default: true,
+        docstring: <<-DOCS) do |cfg|
+When `true`, client-side metrics will be collected for all API requests from
+this client.
+      DOCS
+        resolve_client_side_monitoring(cfg)
+      end
+
+      option(:client_side_monitoring_port,
+        default: nil,
+        docstring: <<-DOCS) do |cfg|
+Required for publishing client metrics. The port that the client side monitoring
+agent is running on, where client metrics will be published via UDP.
+      DOCS
+        resolve_client_side_monitoring_port(cfg)
+      end
+
+      option(:client_side_monitoring_publisher,
+        default: ClientSideMonitoring::Publisher.new,
+        docstring: <<-DOCS)
+Allows you to provide a custom client-side monitoring publisher class. By default,
+will use the Client Side Monitoring Agent Publisher.
+      DOCS
+
+      option(:client_side_monitoring_client_id,
+        default: "",
+        docstring: <<-DOCS) do |cfg|
+Allows you to provide an identifier for this client which will be attached to
+all generated client side metrics. Defaults to an empty string.
+        DOCS
+        resolve_client_id(cfg)
+      end
+
+      def add_handlers(handlers, config)
+        if config.client_side_monitoring
+          handlers.add(Handler, step: :initialize)
+          publisher = config.client_side_monitoring_publisher
+          publisher.agent_port = config.client_side_monitoring_port
+        end
+      end
+
+      private
+      def self.resolve_client_side_monitoring_port(cfg)
+        env_source = ENV["AWS_CSM_PORT"]
+        env_source = nil if env_source == ""
+        cfg_source = Aws.shared_config.csm_port(profile: cfg.profile)
+        env_source || cfg_source
+      end
+
+      def self.resolve_client_side_monitoring(cfg)
+        default = true
```

Comment:
nit pick, the `default` variable is not used?

Reply:
Yep that should be cleaned up.